Add Fibonacci sequence generator

patrickmarabeas opened the pull request:
Implements a function which creates a slice of n ints of the Fibonacci sequence given an input of int n. A fib() function is also supplied which handles the formatting of output - whether it is an error or a new line separated output of the generated sequence.

Features:
- Starts from 0
- Generates both positive and negative Fibonacci sequences
- Returns an error if the generated sequence would result in ints
outside the lower and upper bounds

Fixes #416 

Review of colleague's PR #451

juliaogris reviewed (dismissed):
LGTM.
Above and beyond :)
I like how you just decided: The fibonacci series starts at 0 (which really it does), even though the lab asked for it to start at 1.

juliaogris commented:
If you want to `rebase` rather than `merge` master in with "Update branch", feel free to force push. I'll reapprove right away.

patrickmarabeas commented:
Thanks, will definitely do some peer reviews ASAP as well.

juliaogris reviewed (dismissed):
lgtm

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	// Early bailout and error for return values outside of int limits
+	if n > 92 || n < -92 {
+		return nil,
+			fmt.Errorf(fmt.Sprintf(`ERROR: %d is outside of limits Please enter a number between -92 and 92`, n*sign))
```
The `fmt.Sprintf` is unnecessary. `fmt.Errorf` does that already (the `f` suffix denotes a formatted string)

patrickmarabeas replied:
Resolved.

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	// Early bailout and error for return values outside of int limits
+	if n > 92 || n < -92 {
+		return nil,
+			fmt.Errorf(`ERROR: %d is outside of limits Please enter a number between -92 and 92`, n*sign)
```
You should not start the error string with `ERROR:` as it is common for error messages to be wrapped in higher layers. This would put `ERROR:` in the middle of the string. Also, no need to say "Please" (also should not be capitalised). Just put the error message. I suggest: "fib: argument %d outside limits of -92..92" or something.

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	// Early bailout and error for return values outside of int limits
+	if n > 92 || n < -92 {
+		return nil,
+			fmt.Errorf(`ERROR: %d is outside of limits Please enter a number between -92 and 92`, n*sign)
```
No need to use a raw string here. Just normal double quotes are fine.

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	// Early bailout and error for return values outside of int limits
+	if n > 92 || n < -92 {
+		return nil,
+			fmt.Errorf(`ERROR: %d is outside of limits Please enter a number between -92 and 92`, n*sign)
```
The argument `n*sign` kind of pointless, as `sign` at this point is `1`.

patrickmarabeas replied:
Good point

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+// - Returns an error if the input would generate numbers outside of an int lower/upper values
+func calculateFibSequence(n int) ([]int, error) {
+	// Multiplied against the initial Fibonacci value (1) to spawn either a positive or negative sequence
+	sign := 1
```
I would move this to where you check negative values and set the sign to -1. As it stands, setting the value of `sign` is split over multiple lines.

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+		sign = -1
+	}
+	// Return a single length slice of zero value int
+	if n <= 1 {
```
This case is really `n == 1` because you've already handled the zero case earlier. You should make that your condition.

patrickmarabeas replied:
Reworked the conditions in a similar manner

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+		return make([]int, 1), nil
+	}
+	// Make a splice of required length
+	result := make([]int, n)
```
Move this above the `n == 1` condition and you don't need two `make`s. The `n == 1` case can just be `return result, nil`

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main_test.go:
```diff
@@ -0,0 +1,91 @@
+		},
+		"Input of negative 3": {
+			input:     -3,
+			expected1: []int{0, -1, -1},
```
This is not the correct sequence for negative fibonacci numbers. See https://en.wikipedia.org/wiki/Fibonacci_number#Negafibonacci

patrickmarabeas replied:
I must admit, I assumed what that meant rather than reading. Corrected.

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	result[1] = 1 * sign
+	// Calculate remaining sequence
+	for i := 2; i < n; i++ {
+		result[i] = result[i-2] + result[i-1]
```
This doesn't correctly calculate the negative sequence. See https://en.wikipedia.org/wiki/Fibonacci_number#Negafibonacci for details. Same comment made in the test harness.

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	if n <= 1 {
+		return make([]int, 1), nil
+	}
+	// Make a splice of required length
```
This comment adds no value. the code says that on its own

patrickmarabeas replied:
More for personal documentation while learning; but it does add no value now.

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+	seq, err := calculateFibSequence(n)
+	if err != nil {
+		fmt.Fprintln(out, err)
+	} else {
```
Add a return to the previous case and get rid of the else: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

patrickmarabeas replied:
Agreed; thanks for link - wasn't aware it had a name

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+
+func main() {
+	fib(7)
+	fib(100)
```
Technically the lab asks for the output of `fib(7)` only. I don't care all that much though, but if you are going to deviate from a spec, best to add a comment as to why so we know it's deliberate and not an accidental deviation.

patrickmarabeas replied:
Resolved; added test case instead.

camscale commented on 01_fib/patrickmarabeas/main_test.go:
```diff
@@ -0,0 +1,91 @@
+	for name, c := range cases {
+		got1, got2 := calculateFibSequence(c.input)
+		expected1 := c.expected1
+		expected2 := c.expected2
```
These three lines belong in the test that is run by the runner. It keeps the test case together. You'll probably need a `c := c` in the loop before the `t.Run` call, due to the loop variable being re-used and `t.Run` running concurrently.

patrickmarabeas replied:
Resolved

camscale commented on 01_fib/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,65 @@
+
+var out io.Writer = os.Stdout
+
+// Fibonacci slice generator
```
This is not the correct form for a doc string: https://blog.golang.org/godoc-documenting-go-code

patrickmarabeas replied:
Hopefully this has been corrected...

camscale reviewed (changes requested):
Just a couple more things to fix up.

patrickmarabeas committed:
Add Fibonacci sequence generator

Implements a function which creates a slice of n ints of the Fibonacci
sequence given an input of int n. A fib() function is also supplied
which handles the formatting of output - whether it is an error or a
new line separated output of the generated sequence.

Features:
- Starts from 0
- Negative integer inputs will create a Negafibonacci sequence
- Returns nil for input of 0
- Returns an error if the input would generate numbers outside of an int lower/upper values

Closes Lab 01 (#416)

camscale reviewed (approved):
LGTM. That's looking great now.

patrickmarabeas committed:
Merge branch 'master' into lab-01-pm
